hs/CalendarResponse.hs

```diff
@@ -0,0 +1,284 @@
+
+-- | Creates an event for each start/end time.
+eventsByDate :: String -- ^ Course code.
+             -> String -- ^ Course section.
+             -> String -- ^ Current system time.
+             -> [String] -- ^ Holidays dates.
+             -> ([String], -- ^ Start time
+                 [String], -- ^ End time
+                 (String, String)) -- ^ Start/End date.
+             -> [String]
+eventsByDate code sect systemTime holidays (start, end, dates) =
+    concatMap (eventsStr code sect systemTime dates holidays) (zip start end)
+
+-- | Generates the string that represents each event.
+eventsStr :: String -- ^ Course code.
+          -> String -- ^ Course section.
+          -> String -- ^ Current system time.
+          -> (String, String) -- ^ Start/End date.
+          -> [String] -- ^ Holidays dates.
+          -> (String, String) -- ^ Start/End time.
+          -> [String]
+eventsStr code sect systemTime (startDate, endDate) holidays (start, end)
+    | (startDate == "" || endDate == "") = []
+    | otherwise =
+        ["BEGIN:VEVENT",
+         "DTSTAMP:" ++ systemTime,
+         "DTSTART;TZID=America/Toronto:" ++ startDate ++ start,
+         "DTEND;TZID=America/Toronto:" ++ startDate ++ end,
+         "RRULE:FREQ=WEEKLY;UNTIL=" ++ endDate ++ "000000Z"] ++
+        map (\date -> "EXDATE;TZID=America/Toronto:" ++ date ++ start)
+            holidays ++
+        ["ORGANIZER:University of Toronto",
+         "SUMMARY:" ++ code ++ " " ++ sect,
+         "CATEGORIES:EDUCATION",
+         "END:VEVENT"]
+
+-- ** Ordering data
+
+-- | Orders the time fields for each course; which were obtained from the
+-- database
+orderData :: [Time] -> [[[Double]]]
+orderData courseFields = filter (not . null)
+                                (assignDay $ map timeField courseFields)
+
+-- | Organizes the data by the day of the week, assigning a position based on
+-- the day in the time field.
+assignDay :: [[Double]] -> [[[Double]]]
+assignDay lst = groupBy (\x y -> head x == head y) sortedList
+    where
+        sortedList = sortBy compare lst
+
+-- ** Start time
+
+-- | Obtains the start times for each course by day.
+startTime :: [[[Double]]] -> [[String]]
+startTime dataInOrder =
+    map (\dataByDay -> map getStrTime
+                           (head (sortList dataByDay) :
+                                   getStartConsecutives (sortList dataByDay)))
+        dataInOrder
+    where
+        sortList dataDay = sort $ map (!! 1) dataDay
+
+-- | Gets the start times that are not the very first start time.
+getStartConsecutives :: [Double] -> [Double]
+getStartConsecutives listTimes =
+    filter (/= 30.0) [if listTimes !! i == listTimes !! (i + 1) - 0.5
+                      then 30.0
+                      else listTimes !! (i + 1)| i <- [0 .. lenForComparison]]
+    where
+        lenForComparison = length listTimes - 2
+
+-- ** End time
+
+-- | Obtains the end times for each course by day.
+endTime :: [[[Double]]] -> [[String]]
```

Comment:
[[[Double]]] is a list containing the time fields ordered by day. Each [[Double]] is a list that contains the time fields for just one day of the week. endTime :: [[[Double]]] -> [[String]] could be replaced by endTime :: TimeFieldsByDay -> EndTimesByDay